File: codes/B03_truss/matlab/run_LinStaticRod_template_v10.m
% BK10A6400 Basics of FE Analysis (FEMBasics2025autumn)
% Teacher in charge: Marko Matikainen (MKM)
% Code template for the Bonus Task 3.
%
% Goal: The code solves global displacements at node 2 and member forces of a simple rod structure.
% ################## Have fun with FE !!!! #########################
% Coded by MKM for students' use in the FEMBasics2025 course

% Node and elemental numbering (ID) for a structure:
%           o N1
%            \
%       Elem1 \   
%              o  N2  
%       Elem2 /|
%            /  F 
%           o N3


% First matlab-based FE code for a simple truss structure
clear all;
close all;
clc;
format shortG;  % just to display values in more convient manner

% Units are mm and N and radians
% Rod 1
E1=210000;
A1=100;
L1=1000;
alpha1=??;          % TIP: Give the correct angle (in radians)!!!!

% Rod 2
E2=210000;
A2=50;
L2=1000;
alpha2=??;`         % TIP: Give the correct angle (in radians)!!!!

% Force (in vertical (global) direction at node 2)
F=-100000;


% Transformation into a global coordinate system
% Transformation matrix for a rod 1 
T1=TRod(alpha1);
% Transformation matrix for a rod 2
T2=TRod(alpha2);

% Local elemental stiffness matrices
Kloc1=KlocRod(E1,A1,L1);
Kloc2=KlocRod(E2,A2,L2);

% Global elemental stiffness matrix for a rod 1
Kglob1=T1'*Kloc1*T1;
% Global elemental stiffness matrix for a rod 2
Kglob2=T2'*Kloc2*T2;

% Global stiffness matrix assembling for a whole structure (6x6)
Kglob=zeros(6,6);           % Initialize zero matrix
                            % TIP: make assembling!


% Load vector (in a global coordinate systems)
fglob=zeros(6,1);           % Initialization
fglob(??)=??;                 % TIP: Substitute nodal force with respectt to a correct DOF !

% Boundary conditions (in a global coordinate systems)
% After bc's, global stiffness matrix is 2x2 and external force vector is
% 2x1.
% Boundary conditions ???? are fixed so corresponding rows and
% columns are eliminated from Kglob and fglob. 
Kglobc=Kglob(?,?);      % TIP: take correct (free unconstrained) components in Klobc
fglobc=fglob(?:?);      % TIP: take correct (free unconstrained) components in fglobc 

% Now it is good to check that the stiffness matrix has full rank and/or det
% ~= 0 to be sure that it is invertible.
rank(Kglobc);
det(Kglobc);

% Solve nodal displacement (in a global coordinate systems)
uglobc=Kglobc\fglobc;

% Let's collect all nodal displacements (solved and fixed)
ugloball=zeros(6,1);
ugloball(??:??)=uglobc;       % TIP: you need to add displacements of node 2 in correct place in vector ugloball

% Makes elemental nodal displacement vectors
uglob1=ugloball(??:??);         % TIP: What are nodal displacements IDs of rod 1...jus substitute those in uglob1 
uglob2=ugloball(??:??);         % TIP: What are nodal displacements IDs of rod 2...jus substitute those in uglob2

% Solve element's axial displacements (in a local coordinate systems)
uloc1=??;           % TIP: make transformation from glob to loc.
uloc2=??;           % TIP: make transformation from glob to loc.

% Let's solve member forces f=K u
floc1=??;           % TIP: calculate member forces for rod 1    
floc2=??;           % TIP: calculate member forces for rod 2

% Next ones are asked to return in Moodle
disp('Determinant of Kglob')
det(Kglob)             % Determinant of unconstrained global stiff matrix (should be 0 or really close)
disp('Determinant of Kglobc')
det(Kglobc)            % Determinant of constrained global stiff matrix (>>0)
disp('Displacement at node 2')
ugloball(3:4)          % Displacements at node 2
disp('Local elemental displacements')
uloc1
uloc2
disp('Member forces')
floc1
floc2

File: bonustasks/BonusTask?/run_LinStaticRod_template_v10.py
import numpy as np
from KlocRod import kloc_rod
from TRod import t_rod


# Node and elemental numbering (ID) for a structure:
#           o N1
#            \
#       Elem1 \   
#              o  N2  
#       Elem2 /|
#            /  F 
#           o N3

# Units: mm and N; angles in radians
# Rod 1
E1 = 210000.0
A1 = 100.0
L1 = 1000.0
alpha1 = ??

# Rod 2
E2 = 210000.0
A2 = 50.0
L2 = 1000.0
alpha2 = ??

# Force: vertical at node 2 (negative y is downward)
F = -100000.0

# Transformations
T1 = t_rod(alpha1)
T2 = t_rod(alpha2)

# Local stiffness
Kloc1 = kloc_rod(E1, A1, L1)
Kloc2 = kloc_rod(E2, A2, L2)

# Global element stiffness
Kglob1 = T1.T @ Kloc1 @ T1
Kglob2 = T2.T @ Kloc2 @ T2

# Assemble global 6×6 (DOF: [u1, v1, u2, v2, u3, v3])
Kglob = np.zeros((6, 6), dtype=float)
Kglob[0:4, 0:4] = Kglob1          # elem 1: nodes 1–2
Kglob[2:6, 2:6] = Kglob[2:6, 2:6]+ Kglob2          # elem 2: nodes 2–3

# Global load vector
fglob = np.zeros((6,1),dtype=float)
fglob[?] = ??  # at v2

# Apply BCs: u1=v1=u3=v3=0 -> free DOFs: [u2, v2] -> indices [2,4]
Kglobc = Kglob[?,?]
fglobc = fglob[?:?]

# Checks
det_full = np.linalg.det(Kglob)
det_red = np.linalg.det(Kglobc)
rank_red = np.linalg.matrix_rank(Kglobc)

# Solve reduced system
uglobc = np.linalg.solve(Kglobc, fglobc)

# Collect all displacements
ugloball = np.zeros((6,1),dtype=float)
ugloball[?:?] = uglobc

# Elemental global displacement vectors
uglob1 = ugloball[?:?]   # [u1, v1, u2, v2]
uglob2 = ugloball[?:?]   # [u2, v2, u3, v3]

# Transform to local
uloc1 = ??
uloc2 = ??

# Local member forces f = Kloc * u_loc
floc1 = ??
floc2 = ??

# Output (mirrors MATLAB prints)
print('Determinant of Kglob')
print(det_full:.6g)
print('Determinant of Kglobc')
print(det_red:.6g)
print('Displacement at node 2')
print(ugloball[2:4])
print('Local elemental displacements')
print('uloc1 =', uloc1)
print('uloc2 =', uloc2)
print('Member forces (N)')
print('floc1 =', floc1)
print('floc2 =', floc2)